R/init.R

```diff
@@ -14,706 +29,81 @@ if (interactive() &&
-      show_globalenv <- isTRUE(getOption("vsc.globalenv", TRUE))
-      if (show_globalenv) {
-        globalenv_file <- file.path(dir_session, "globalenv.json")
-        globalenv_lock_file <- file.path(dir_session, "globalenv.lock")
-        file.create(globalenv_lock_file, showWarnings = FALSE)
-
-        update_globalenv <- function(...) {
-          tryCatch({
-            objs <- inspect_env(.GlobalEnv, globalenv_cache)
-            jsonlite::write_json(objs, globalenv_file, force = TRUE, pretty = FALSE)
-            cat(get_timestamp(), file = globalenv_lock_file)
-          }, error = message)
-          TRUE
-        }
-
-        update_globalenv()
-        addTaskCallback(update_globalenv, name = "vsc.globalenv")
-      }
-
-      removeTaskCallback("vsc.plot")
-      show_plot <- !identical(getOption("vsc.plot", "Two"), FALSE)
-      if (show_plot) {
-        dir_plot_history <- file.path(dir_session, "images")
-        dir.create(dir_plot_history, showWarnings = FALSE, recursive = TRUE)
-        plot_file <- file.path(dir_session, "plot.png")
-        plot_lock_file <- file.path(dir_session, "plot.lock")
-        file.create(plot_file, plot_lock_file, showWarnings = FALSE)
-
-        plot_history_file <- NULL
-        plot_updated <- FALSE
-        null_dev_id <- c(pdf = 2L)
-        null_dev_size <- c(7 + pi, 7 + pi)
-
-        check_null_dev <- function() {
-          identical(dev.cur(), null_dev_id) &&
-            identical(dev.size(), null_dev_size)
-        }
-
-        new_plot <- function() {
-          if (check_null_dev()) {
-            plot_history_file <<- file.path(dir_plot_history,
-              format(Sys.time(), "%Y%m%d-%H%M%OS6.png"))
-            plot_updated <<- TRUE
-          }
-        }
-
-        options(
-          device = function(...) {
-            pdf(NULL,
-              width = null_dev_size[[1L]],
-              height = null_dev_size[[2L]],
-              bg = "white")
-            dev.control(displaylist = "enable")
-          }
-        )
-
-        update_plot <- function(...) {
-          tryCatch({
-            if (plot_updated && check_null_dev()) {
-              plot_updated <<- FALSE
-              record <- recordPlot()
-              if (length(record[[1L]])) {
-                dev_args <- getOption("vsc.dev.args")
-                do.call(png, c(list(filename = plot_file), dev_args))
-                on.exit({
-                  dev.off()
-                  cat(get_timestamp(), file = plot_lock_file)
-                  if (!is.null(plot_history_file)) {
-                    file.copy(plot_file, plot_history_file, overwrite = TRUE)
-                  }
-                })
-                replayPlot(record)
-              }
-            }
-          }, error = message)
-          TRUE
-        }
-
-        setHook("plot.new", new_plot, "replace")
-        setHook("grid.newpage", new_plot, "replace")
-
-        rebind(".External.graphics", function(...) {
```

Comment:
@ManuelHentschel   This change no longer forces the replacing of `base::.External.graphics` to a hacked version to trigger plot replay.

Reply:
Should we just revert this to the old behavior or do you know a different fix? I tried to avoid messing with the `base` namespace itself, but if it's the only (feasible) way to achieve this, it of course makes sense to make an exception

Reply:
Yes, we rely on this replacement to make it possible for us to make all plot functions including e.g. `abline()` that adds graphics elements to the existing plot to trigger plot replay so that the plot is saved to the png file we watch in vscode.

Reply:
That's about the only way I know so that we don't have to write a full graphics device to know when the plot is updated. Therefore, I revert this via #625.